[ivan24004] iP

ivan24004 opened the pull request:
### Remi
> "If you are what you eat, then I only want to eat the good stuff" - Remy (Ratatouille)

Remi frees your mind of keeping track of all your tasks. It's:
- Text-based
- Named after my cat 🐱
- FREE ~~forever~~ _(for now)_

Features:
- [x] Managing tasks
- [ ] Managing deadlines (in development)
- [ ] Reminders (in development)

How do I download this magnificent project?
1. download it from [here](https://github.com/ivan24004/ip/releases/tag/A-Jar)
2. double click it
3. add your tasks
4. let it manage tasks for you ☺

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Remi {
    public static void main(String[] args) {
        Ui chatbot = new Ui();
        chatbot.doIoLoop();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

ivan24004 committed:
Created output classes

ivan24004 committed:
Created main logic class of the Chatbot

ivan24004 committed:
made data field final

ivan24004 committed:
Renamed Duke chat bot to Remi chat bot

ivan24004 committed:
Renamed Duke chat bot to Remi chat bot

ivan24004 committed:
Renamed Duke chat bot to Remi chat bot

ivan24004 committed:
Added an inputter static class

ivan24004 committed:
Changed chat bot to a static class

ivan24004 committed:
Added javadocs to the Outputter class

ivan24004 committed:
made scanner final

ivan24004 committed:
moved chatbot to diff folder for better package structure

ivan24004 committed:
Enforced camel case on the ioLoop function

ivan24004 committed:
Added task and tasklist classes

ivan24004 committed:
Added command and command list classes to model the different functionalities of the chatbot.

ivan24004 committed:
made chatbot non static

ivan24004 committed:
moved commands responsibility to the CommandList object

ivan24004 committed:
finished functionality for task2

ivan24004 committed:
modified Task and TaskList to add marking and unmarking tasks

ivan24004 committed:
finished adding the mark and unmark commands

ivan24004 committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	src/main/java/Chatbot.java

ivan24004 committed:
added a note that multiple string concats may be slow

ivan24004 committed:
removed unnecessary imports

ivan24004 committed:
added a todo class

ivan24004 committed:
added a deadline class

ivan24004 committed:
modified the name of the variable to fit convention

ivan24004 committed:
created an event class

ivan24004 committed:
removed trailing newline in toString method

ivan24004 committed:
added the todo, deadline, and event commands

ivan24004 committed:
added the size functionality to the taskList

ivan24004 committed:
added the size functionality to the CommandList

ivan24004 committed:
added a use case scenario to the inputs file

ivan24004 committed:
made the output messages have dashed lines on the top and bottom

ivan24004 committed:
removed trailing new line in welcome message

ivan24004 committed:
renamed the test from Duke to Remi

ivan24004 committed:
modified to remove an emoji in the message after unmarking a task

ivan24004 committed:
test cases are now written

ivan24004 committed:
created a custom exception class

ivan24004 committed:
indicated that parseAndRun may throw an error

ivan24004 committed:
added a try catch block to catch errors

ivan24004 committed:
declared that the Command functional interface can throw a RemiError

ivan24004 committed:
added error handling to the different commands

ivan24004 committed:
added the ability to remove tasks

ivan24004 committed:
forgot semicolon

ivan24004 committed:
added the delete functionality

ivan24004 committed:
Add Storage class

ivan24004 committed:
Add StoredTaskList class that automatically stores changes in the TaskList

ivan24004 committed:
Added Storage class for interacting with the storage

ivan24004 committed:
Modify the Chatbot to use StoredTaskList

ivan24004 committed:
Modify the Chatbot to use StoredTaskList

ivan24004 committed:
Catch the IOException and print it out

ivan24004 committed:
Added more parsable string versions of the Task objects

ivan24004 committed:
Close the output file after writing

ivan24004 committed:
Convert the toString method to use StringBuilders

ivan24004 committed:
Change delimiter from , to |

ivan24004 committed:
Parse the storage file

ivan24004 committed:
Finish the get() method

ivan24004 committed:
Use the get() method when creating the StoredTaskList

ivan24004 committed:
Debug the isDone item

ivan24004 committed:
Add tasks.txt to .gitignore

ivan24004 committed:
Add formatDate() as a protected method

ivan24004 committed:
Use the formatDate() method

ivan24004 committed:
Change the name of Chatbot to Ui

ivan24004 committed:
Move the files into the Remi package

ivan24004 committed:
Modify the test files to call remi.Remi instead of Remi

ivan24004 committed:
Merge branch 'add-gradle-support'

ivan24004 committed:
Change the Ui to say "Hi I'm Remi" instead of "remi.Remi"

ivan24004 committed:
Changed duke to remi

ivan24004 committed:
Remove seedU from mainClass

ivan24004 committed:
Add new CommandListTest junit class

ivan24004 committed:
Expand the imports

ivan24004 committed:
Add test cases for CommandList.runKeyword

ivan24004 committed:
Avoid using wildcard imports

ivan24004 committed:
Add tests for parsableString todo

ivan24004 committed:
Add jar file creation

ivan24004 committed:
Remove hard-coded BORDER_LENGTH value and put it into a constant variable

ivan24004 committed:
Change parsableString to getParsableString to comply with coding standard's requirement of verb usage

ivan24004 committed:
Change parsableString to getParsableString to comply with coding standard's requirement of verb usage

ivan24004 committed:
Change ioLoop to doIoLoop to comply with coding standard's requirement of verb usage

ivan24004 committed:
Add javadoc for Ui.ioLoop

ivan24004 committed:
Add javadoc for Ui.exitIoLoop

ivan24004 committed:
Add javadoc for TaskList public methods

ivan24004 committed:
Add javadoc for the different Task classes and their public methods.

ivan24004 committed:
Add class descriptor and remove redundant JavaDoc

ivan24004 committed:
Add class descriptor

ivan24004 committed:
Add the checkQuery method

ivan24004 committed:
Add the getTask method that queries using a keyword

ivan24004 committed:
Add the find command

ivan24004 committed:
Merge branch 'branch-A-CodingStandard'

ivan24004 committed:
Merge the javadoc branch

bachletuankhai commented on src/main/java/remi/model/Deadline.java:
```diff
@@ -0,0 +1,34 @@
+ * Represents a deadline task that ends by a certain time. Subclass of Task.
+ */
+public class Deadline extends Task{
+    private String by;
```
Perhaps it is better to give the due date variable a meaningful and self-explaining name? Something like `dueDate`.

bachletuankhai commented on src/main/java/remi/model/TaskList.java:
```diff
@@ -0,0 +1,120 @@
+ * Stores the list of Tasks, allows you to add, remove, and find tasks by index.
+ */
+public class TaskList {
+    private ArrayList<Task> taskList;
```
The recommended naming for collection is by using plural noun. But I am not sure this is bad 🤔.

bachletuankhai commented on src/main/java/remi/model/Ui.java:
```diff
@@ -0,0 +1,57 @@
+ */
+public class Ui {
+
+    private boolean exitLoop;
```
boolean variables should be named to sound like boolean. (see [coding standard](https://se-education.org/guides/conventions/java/intermediate.html#naming))

felixchanyy replied:
Should the boolean variables sound like booleans like isExit?

bachletuankhai commented on src/main/java/remi/parsing/Parser.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class Parser {
+
+    CommandList commandList;
+    private String[] getKeywordArgsSplit(Message input) {
+        int idx = input.getMessage().indexOf(' ');
+        String[] res = new String[2];
+
+        // this indicates a single word command
+        if (idx == -1) {
+            res[0] = input.getMessage();
+            res[1] = "";
+        } else {
+            res[0] = input.getMessage().substring(0, idx);
+            res[1] = input.getMessage().substring(idx + 1);
+        }
```
Perhaps using `String::split()` would be more convenient here?

bachletuankhai commented on src/main/java/remi/storage/Storage.java:
```diff
@@ -0,0 +1,105 @@
+import remi.model.StoredTaskList;
+import remi.model.ToDo;
+
+import java.io.*;
```
The coding standard is strictly against wildcard imports. I don't see any reason to import everything too. Perhaps you should replace it with named imports instead? And maybe you should try configuring IntelliJ not to merge your name imports into wildcard import.

felixchanyy replied:
Should this be listed explicitly?

bachletuankhai commented on src/test/java/remi/model/commands/CommandListTest.java:
```diff
@@ -0,0 +1,29 @@
+
+public class CommandListTest {
+    @Test
+    public void addTodo_NoError() {
```
I think the test case naming should follow the format `nameOfMethodTested_inputDescription_expectedOutput`. You are missing the method name here.

felixchanyy replied:
Should the test method names uses the following three-part format featureUnderTest_testScenario_expectedBehavior()?

bachletuankhai commented on src/test/java/remi/model/commands/ToDoTest.java:
```diff
@@ -0,0 +1,15 @@
+public class ToDoTest {
+
+    @Test
+    public void parsableString_dummyTask_returnString() {
```
I like how the test case's name fully describes the test case here.

bachletuankhai reviewed (commented):
LGTM. I like how you organised your classes, package names and class design. Keep up the good work!

ivan24004 committed:
Make simple javafx input box.

ivan24004 committed:
Rename to GuiLauncher

ivan24004 committed:
Add gui elements for basic messaging app.

ivan24004 committed:
Add profile picture images.

ivan24004 committed:
Flip the messages of the chatbot.

felixchanyy commented on src/main/java/remi/model/TaskList.java:
```diff
@@ -0,0 +1,120 @@
+     *
+     * @param t task to be added
+     */
+    public void addTask(Task t) {
```
I like how methods are verbs and written in camelCase.

felixchanyy commented on src/main/java/remi/io/Outputter.java:
```diff
@@ -0,0 +1,19 @@
+ */
+public class Outputter {
+
+    private static final int BORDER_LENGTH = 50;
```
I like the usage of SCREAMING_SNAKE_CASE

felixchanyy commented on src/main/java/remi/model/Ui.java:
```diff
@@ -0,0 +1,57 @@
+        Outputter.outputMessage(new Message("Hello! I'm Remi\n" + "What can I do for you?"));
+        Storage.get();
+
+        while (!exitLoop) {
```
I like how Egyptian style is being used

ivan24004 committed:
Seperate the loop into parts. Allow messages to be directly sent to the parser instead of having to use an Inputter and Outputter based on stdio.

felixchanyy commented on src/main/java/remi/storage/Storage.java:
```diff
@@ -0,0 +1,105 @@
+            System.out.println(err.toString());
+        }
+
+        boolean corruptedFile = false;
```
I noticed the same issue in other places too.

felixchanyy reviewed (commented):
I like how clean this code is. Just some nits to fix.

ivan24004 committed:
Add fxml files.

ivan24004 committed:
Remove indent.

ivan24004 committed:
Modify the controller class.

ivan24004 committed:
Add controller class.

ivan24004 committed:
Make changes to use fxml.

ivan24004 committed:
Modify exitLoop local variable to getIsFinished.

ivan24004 committed:
Reorder imports

ivan24004 committed:
Add description in the javadocs

ivan24004 committed:
Add the checkstyle module

ivan24004 committed:
Fix import order

ivan24004 committed:
Add the checkstyle xml files

These files were taken from the ab3 project repository. It follows the
checkstyle format. It can be used in most IDEs to lint and provide
static analysis.

ivan24004 committed:
Assert that the option and input are non-empty

ivan24004 committed:
Shorten the parseData function by separating it

ivan24004 committed:
Avoid unnecessary nesting

ivan24004 committed:
Modify import order

ivan24004 committed:
Add spacing to improve readability

ivan24004 committed:
Change spacing

ivan24004 committed:
Change spacing to make separation clearer

ivan24004 committed:
Remove unnecessary use of this

ivan24004 committed:
Merge pull request #2 from ivan24004/branch-A-Assertions

Add new asserts

ivan24004 committed:
Merge pull request #1 from ivan24004/branch-A-CodeQuality

Code quality improvements

ivan24004 committed:
Add sortlist command to view task list sorted alphabetically

ivan24004 committed:
Add getLabel method

ivan24004 committed:
Add list sorting functionality to the TaskList

ivan24004 committed:
Allow messages longer than 5 lines to be displayed

ivan24004 committed:
Change the exit command to work with javafx

ivan24004 committed:
Remove unused variable

ivan24004 committed:
Fix comment spacing issues

ivan24004 committed:
Fix style issues.

ivan24004 committed:
Split the CommandList class into multiple Command classes

ivan24004 committed:
Fix import spacing

ivan24004 committed:
Add the remi title to the GUI

ivan24004 committed:
Add a get initial message function

ivan24004 committed:
Make the getInitialMessage method static

ivan24004 committed:
Show an initial message upon initialization

ivan24004 committed:
Create user guide

ivan24004 committed:
Create user guide

ivan24004 committed:
Add Ui image

ivan24004 committed:
Update link in README, original link was wrong
